src/mgxs_data.F90

```diff
@@ -172,28 +167,120 @@ end subroutine read_mgxs
+! material appears at in the model.
+!===============================================================================
+
+  subroutine get_mat_kTs(kTs)
+    type(VectorReal), allocatable, intent(out) :: kTs(:)
+
+    integer :: i, j
+    integer :: i_material  ! Index in materials array
+    real(8) :: kT ! temperature in MeV
+
+    allocate(kTs(size(materials)))
+
+    do i = 1, size(cells)
+      do j = 1, size(cells(i) % material)
+        ! Skip any non-material cells and void materials
+        if (cells(i) % material(j) == NONE .or. &
+             cells(i) % material(j) == MATERIAL_VOID) cycle
+
+        ! Get temperature of cell (rounding to nearest integer)
+        if (size(cells(i) % sqrtkT) > 1) then
+          kT = cells(i) % sqrtkT(j)**2
+        else
+          kT = cells(i) % sqrtkT(1)**2
+        end if
+
+        i_material = material_dict % get_key(cells(i) % material(j))
+
+        ! Add temperature if it hasn't already been added
+        if (find(kTs(i_material), kT) == -1) then
+          call kTs(i_material) % push_back(kT)
+        end if
+
+      end do
+    end do
+
+  end subroutine get_mat_kTs
+
+!===============================================================================
+! GET_TEMPERATURES returns a list of temperatures that each MGXS table
+! appears at in the model. Later, this list is used to determine the actual
+! temperatures to read (which may be different if interpolation is used)
+!===============================================================================
+
+  subroutine get_temperatures(temps)
+    type(VectorReal), allocatable, intent(out) :: temps(:)
+
+    integer :: i, j, k
+    integer :: i_nuclide    ! index in nuclides array
+    integer :: i_material
+    real(8) :: temperature  ! temperature in Kelvin
+
+    allocate(temps(n_nuclides_total))
+
+    do i = 1, size(cells)
+      do j = 1, size(cells(i) % material)
+        ! Skip any non-material cells and void materials
+        if (cells(i) % material(j) == NONE .or. &
+             cells(i) % material(j) == MATERIAL_VOID) cycle
+
+        ! Get temperature of cell (rounding to nearest integer)
+        if (size(cells(i) % sqrtkT) > 1) then
+          temperature = cells(i) % sqrtkT(j)**2 / K_BOLTZMANN
+        else
+          temperature = cells(i) % sqrtkT(1)**2 / K_BOLTZMANN
+        end if
+
+        i_material = material_dict % get_key(cells(i) % material(j))
+        associate (mat => materials(i_material))
+          NUC_NAMES_LOOP: do k = 1, size(mat % names)
+            ! Get index in temps array
+            i_nuclide = nuclide_dict % get_key(to_lower(mat % names(k)))
+
+            ! Add temperature if it hasn't already been added
+            if (find(temps(i_nuclide), temperature) == -1) then
+              call temps(i_nuclide) % push_back(temperature)
+            end if
+          end do NUC_NAMES_LOOP
+        end associate
+      end do
+    end do
+
+  end subroutine get_temperatures
```

Comment:
Yep, I made the sab argument in input_xml.F90 optional and then moved this in to geometry_header.F90.  Based on the dependencies it was either that or material_header.F90.